fdk/http_stream.py

```diff
@@ -0,0 +1,121 @@
+        if "application/json" in content_type:
+            return bytes(ujson.dumps(data), "utf8")
+        if "text/plain" in content_type:
+            return bytes(str(data), "utf8")
+        if "application/xml" in content_type:
+            # returns a bytearray
+            if isinstance(data, str):
+                return bytes(data, "utf8")
+            return ElementTree.tostring(data, encoding='utf8', method='xml')
+        if "application/octet-stream" in content_type:
+            return data
+    return
+
+
+def encap_headers(headers, status, content_type):
+    new_headers = hs.GoLikeHeaders({})
+    for k, v in headers.items():
+        if "content_type" not in k:
+            new_headers.set(constants.FN_HTTP_PREFIX + k, v)
+
+    new_headers.set("Fn-Http-Status", str(status))
+    new_headers.set("Content-Type", content_type)
+    return new_headers
+
+
+def handle(handle_func):
+    async def pure_handler(request):
+        log.log("in pure_handler")
+        data = None
+        if (request.has_body and
+                request.content_length is not None and
+                request.content_length > 0):
+            log.log("request comes with data")
+            data = await request.content.read(request.content_length)
+        response = await runner.handle_request(
+            handle_func, None, constants.HTTPSTREAM,
+            request=request, data=data)
+        log.log("request execution completed")
+        headers = (response.headers()
+                   if isinstance(response, rtypes.RawResponse)
+                   else response.headers)
+        response_content_type = headers.get(
+            "content-type", "application/json"
+        )
+        headers.delete("content-type")
+
+        headers = encap_headers(
+            headers, response.status(), response_content_type)
+        kwargs = {
+            "status": response.status(),
+            "headers": headers.http_raw()
+        }
+
+        sdata = serialize_response_data(
+            response.body(), response_content_type)
+
+        if response.status() >= 500:
+            kwargs.update(reason=sdata)
+        else:
+            kwargs.update(body=sdata)
+
+        log.log("sending response back")
+        try:
+            resp = web.Response(**kwargs)
+        except (Exception, BaseException) as ex:
+
+            resp = web.Response(
+                text=str(ex), reason=str(ex),
+                status=500, content_type="text/plain", headers={
+                    "Fn-Http-Status": str(500)
+                })
+
+        return resp
+
+    return pure_handler
+
+
+def setup_unix_server(handle_func, loop=None):
+    log.log("in setup_unix_server")
+    app = web.Application(loop=loop)
+
+    app.router.add_post('/{tail:.*}', handle(handle_func))
+
+    return app
+
+
+def start(handle_func, uds, loop=None):
+    log.log("in http_stream.start")
+    app = setup_unix_server(handle_func, loop=loop)
+    web.run_app(app, path=uds.lstrip("unix:"),
```

Comment:
The web server I use is an event-driven (an event loop-driver server on top of libuv), a socket created with a particular listener (unix socket site listener), one a web server runner gets all sites assigned it starts one major and N minor coroutines, major coroutine pinned to the C EPOLL, so, once the data reach out the socket from Fn side kernel notifies the major coroutine to do protocol-framed dispatching (an http route mapping). So, it's not automatic but not in the way you think, the kernel notifies (through EPOLL) a coroutine about the data being written to a socket.

Reply:
So, this is the safest way to do this with Python. The way it was done for Go FDK, unfortunately, breaks socket operations, I made a gist with an alternative `start` function (see https://gist.github.com/denismakogon/9b7177e1892cd0d7cefcd71c16051315), but it doesn't work because the socket is broken and is not writable/readable on both sides of it.

Reply:
with proper symlink, it works.